Nick's BP: Pinned Agenda Items

nickoferrall opened the pull request:
This PR is for pinning agenda items: Issue #3839 

**Acceptance Criteria**

- [x] When hovering over an agenda item, a "pin" icon appears at the right.
- [x] When hovering the pin icon, a tooltip animates in telling them this will pin the item for their next meeting
- [x] When the pin icon is clicked, the icon changes to the unpin icon.
- [x] The pin/unpin state is propagated to all team members via the team graphql subscription
- [x] When the check-in meeting is ended and the agenda item list is flushed, it is repopulated with all the pinned items

**Changes in this PR**

- Added pin and unpin Material Design icons.
- Added `pinned` to GraphQL types.
- Cloning pinned items and re-adding them once the meeting has ended.
- Updated tooltip with `pointerEvents: 'none'` so that hovering over the tooltip doesn't prevent mouse events from being triggered.
- When clicking an icon, the opacity changes to give the user feedback that the icon has been clicked.
- Removed title attribute from Agenda Items as the native browser tooltip was competing with the implemented tooltip.
- While looking at how other tooltips are implemented in the app, I noticed that there's a small bug in Org Settings - if the user hovers over the disabled button, the tooltip doesn't always close. This is because the disabled button adds an icon to the cursor which means that it's initially outside of the element and doesn't trigger `onMouseLeave`. I've updated this to use `onMouseOver` and `onMouseOut` which bubbles and therefore triggers when the cursor leaves the element or any children (i.e. is the button in this case). 

**Considerations**

- Initially, I wanted to distinguish between the "pinned", "to pin" and "to unpin" state by rotating the pin icon by 45-degrees when it's pinned and using a transition on hover. However, I was concerned that the transition was attracting too much attention to the pins and giving it a Microsoft Powerpoint vibe... so when Terry mentioned that the rotation wasn't his favorite, I removed the transition and 45-degree rotation. I feel that the current approach is a bit cleaner and remains intuitive.
- I feel that it's useful to give the user feedback when clicking the icon so I added a subtle opacity change when it's active. My concern is that this looks a bit like it's flashing and it may be somewhat inconsistent with the overall design as most of the other buttons don't animate on click. I thought I would leave this in though to see what you think.

ackernaut commented:
Hey Nick, there are usually some decent options here for more icon options: https://materialdesignicons.com/ — I typically find what I need and use an SVG at the appropriate MD icon size (18, 24, 36, 48). Depending on how we handle the SVG we could provide it via S3. I noticed that site has a `pin-off` variant that we could consider. Maybe the `pin` turns to `pin-off` on hover to indicate you can remove it.

nickoferrall commented on README.md:
```diff
@@ -72,9 +72,9 @@ For details on all releases, refer to [CHANGELOG.md](./CHANGELOG.md).
-* [jordanh](https://github.com/jordanh)
-* [ackernaut](https://github.com/ackernaut)
-* [mattkrick](https://github.com/mattkrick)
+- [jordanh](https://github.com/jordanh)
```
Look like my prettier settings updated this - I have no strong feelings about `-` instead of `*`!

ackernaut commented:
Hello @nickoferrall — thanks for the latest update. Nice work! Here are a few points of quick feedback:

- In an active meeting, when I tap to pin or un-pin, it also navigates to the item, and I’m not sure this is the right behavior. I’m thinking that the tap/click should pin or un-pin only. I could be wrong

- The `.75s` duration of the rotation feels too slow. I’d recommend somewhere between `200ms` to `400ms` to keep it feeling snappy. The theory there is probably somewhat close to the [Doherty Threshold](https://lawsofux.com/doherty-threshold.html) — a side note: I’m not sure the rotation is my favorite but I think it’s safe to try b/c this feels much more like a personal preference instead of something wrong about it, and I think the rotation provides helpful visual feedback, so +1 for creativity there!

- The alignment for the left nav items is nuanced. I noticed you bumped the inner gutter from `8px` to `4px`. Was there a specific need to change the gutter? We have those controls set up on an 8px grid, otherwise, the alignment is off. The right icons/badges are based on a minimum of `24px` wide and centered in relation to each other. Check out [the rulers on this screenshot in Figma](https://www.figma.com/file/1DSwgYv0nRffpvb0Jnuc6G/Meeting-IA?node-id=190%3A1)

I’ll look at it again with fresh eyes or after an update and let my brain simmer on the UI nuances for a while. Also, unless @mattkrick gets to it sooner, I’ll wait until we get closer to give a more thorough code review myself unless you’re looking for some specific feedback there.

ackernaut commented:
Oh yeah! I forgot to mention a couple of things:
- I really like that the pinned items can be sorted in any order. I like the flexibility 💪 
- As for the `title` attributes that cause native browser tooltips: I think you pointed out that they are competing with the explicit `Tooltip` components — feel free to remove the `title` attributes where it makes sense

nickoferrall commented:
> Oh yeah! I forgot to mention a couple of things:
> 
> * I really like that the pinned items can be sorted in any order. I like the flexibility 💪
> * As for the `title` attributes that cause native browser tooltips: I think you pointed out that they are competing with the explicit `Tooltip` components — feel free to remove the `title` attributes where it makes sense

Thanks for your feedback Terry - the Laws of UX link is really interesting! I've updated my PR with your suggestions. 

Regarding the inner gutter change, I thought that the alignment was off but this was actually because of an update that I made... I've now changed the gutter back to how it was before.

mattkrick commented on packages/server/graphql/mutations/endNewMeeting.ts:
```diff
@@ -86,6 +86,7 @@ const clearAgendaItems = async (teamId: string) => {
   return r
     .table('AgendaItem')
     .getAll(teamId, {index: 'teamId'})
+    .filter({pinned: false})
```
doing this means the tasks created within the agenda item are also persisted.

repro:
1. add an agenda item "watermelon" & pin it
2. add a task "eat it"
3. finish the meeting
4. start a new meeting & navigate to "watermelon". "eat it" is there.

consider cloning the pinned items and re-creating them on meeting end. maybe include a "pinnedParentId" if you like so we know which agenda item it is a clone of (may be useful for analytics later?)

nickoferrall replied:
Thanks for the repro steps @mattkrick . I've added a clone function so the tasks no longer persist. 

However, when `AgendaList` initially renders, the cloned items have the id but the content is undefined. On the second render, the content exists. I've handled this in the filter function in the frontend but I'll try to find out why this is happening.

mattkrick commented on packages/server/graphql/types/AgendaItem.ts:
```diff
@@ -32,6 +32,10 @@ const AgendaItem = new GraphQLObjectType<IAgendaItem, GQLContext>({
     },
+    pinned: {
+      type: new GraphQLNonNull(GraphQLBoolean),
+      description: 'True if the agenda item has been pinned'
```
type is nonnull, but pinned will be null for existing & previous meetings. might want to resolve to false if undefined

nickoferrall replied:
Thanks! Updated

nickoferrall commented on packages/client/modules/teamDashboard/components/AgendaItem/AgendaItem.tsx:
```diff
@@ -82,19 +98,38 @@ const getItemProps = (
   const atmosphere = useAtmosphere()
   const {viewerId} = atmosphere
+  const isHovering = hoveringId === agendaItemId
+  const closedTooltipStatus = 4
```
True - I've made the update 👍

nickoferrall commented on packages/client/modules/teamDashboard/components/AgendaItem/AgendaItem.tsx:
```diff
@@ -109,17 +144,66 @@ const AgendaItem = (props: Props) => {
   }
+
+  const handleMouseMoveItem = () => {
+    // onMouseEnter isn't triggered if the cursor quickly moves over tooltip so check onMouseMove
```
Discussing this one in Slack as changing the portal position didn't resolve this one for me

mattkrick commented on packages/server/graphql/mutations/endNewMeeting.ts:
```diff
@@ -307,7 +357,13 @@ export default {
     const result = await finishMeetingType(completedMeeting, dataLoader)
+
+    pinnedAgendaItems.map((agendaItem) => {
+      clonePinnedAgendaItem(agendaItem, viewerId)
```
this will result in multiple DB hits. can you reduce it to a single bulk insert?

nickoferrall replied:
Thanks - I've updated this 👍

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaList/AgendaList.tsx:
```diff
@@ -86,7 +97,7 @@ const AgendaList = (props: Props) => {
         {(provided) => {
           return (
-            <AgendaListRoot ref={provided.innerRef}>
+            <AgendaListRoot onMouseLeave={handleMouseLeave} ref={provided.innerRef}>
```
from a high level, I'd think the hover state of an item should be encapsulated within the item.
is it necessary to put the hover state on the list?

nickoferrall replied:
Agreed - I reluctantly went for this approach because `onMouseLeave` wasn't being triggered when the cursor went through the tooltip and out of the item. With the `pointerEvent` fix, I've been able to remove this

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaItem/AgendaItem.tsx:
```diff
@@ -109,32 +144,79 @@ const AgendaItem = (props: Props) => {
+          metaContent={
+            <IconBlock
+              onClick={handleIconClick}
+              onMouseMove={handleMouseMoveIcon}
```
i went ahead & tried simplifying the hover & open logic on my own branch. 
i think the problem that you're bumping into is that when the mouse moves from item 1 to 2 it first passes over the tooltip, and then when the tooltip disappears, the mouse lands in the 2nd item without firing an OnMouseEnter event. 

as a hint, if you put a `pointEvents: 'none'` css prop on the tooltip, how can you simplify the hovering logic?

nickoferrall replied:
Well, I'm never going to forget `pointerEvents: 'none'`... Thanks for your suggestion, Matt. It feels great to remove all of the additional complexity

mattkrick commented on packages/client/modules/teamDashboard/components/AgendaItem/AgendaItem.tsx:
```diff
@@ -109,32 +132,61 @@ const AgendaItem = (props: Props) => {
-    </AgendaItemStyles>
+    <>
+      <AgendaItemStyles
+        onMouseEnter={() => setIsHovering(true)}
```
consider onMouseOver instead of onMouseEnter for this edge case:
1. on a team dashboard, have 1 agenda item
2. put your mouse cursor over the agenda item input and click
3. enter the agenda item & hit enter
4. notice that the cursor is now _over_ the 2nd agenda item, but isHovering is false

nickoferrall replied:
Good point! Thanks Matt. I've pushed the update to my branch but I see you beat me to it. Great to see that it's now in master 🎉
